rfcs/generalized-iteration.md

```diff
@@ -0,0 +1,120 @@
+
+```lua
+if getmetatable(gen) and getmetatable(gen).__iter then
+   gen, state, index = getmetatable(state).__iter(gen)
+end
+```
+
+This check is comparatively trivial: usually `gen` is a function, and functions don't have metatables; as such we can simply check the type of `gen` and if it's a table/userdata, we can check if it has a metamethod `__iter`. Due to tag-method cache, this check is also very cheap if the metamethod is absent.
+
+This allows objects to provide a custom function that guides the iteration. Since the function is called once, it is easy to reuse other functions in the implementation, for example here's a node object that exposes iteration through its children:
+
+```lua
+local Node = {}
+Node.__index = Node
+
+function Node.new(children)
+  return setmetatable({ children = children }, Node)
+end
+
+function Node:__iter()
+  return next, self.children
+end
+```
+
+Luau compiler already emits a bytecode instruction, FORGPREP*, to perform initial loop setup - this is where we can evaluate `__iter` as well.
+
+Naturally, this means that if the table has `__iter` metamethod and you need to iterate through the table fields instead of using the provided metamethod, you can't rely on the general iteration scheme and need to use `pairs`. This is similar to other parts of the language, like `t[k]` vs `rawget(t, 'k')`, where the default behavior is overrideable but a library function can help peek behind the curtain.
+
+### Default table iteration
+
+If the argument is a table and it does not implement `__iter` metamethod, we treat this as an attempt to iterate through the table using the builtin iteration order.
+
+> Note: we also check if the table implements `__call`; if it does, we fall back to the default handling. We may be able to remove this check in the future, but we will need this initially to preserve backwards compatibility with custom table-driven iterator objects that implement `__call`. In either case, we will be able to collect detailed analytics about the use of `__call` in iteration, and if neither is present we can emit a specialized error message such as `object X is not iteratable`.
+
+To have a single, unified, iteration scheme over tables regardless of whether they are arrays or dictionaries, we establish the following semantics:
+
+- First, the traversal goes over numeric keys in range `1..k` up until reaching the first `k` such that `t[k] == nil`
+- Then, the traversal goes over the remaining keys, numeric and otherwise, in unspecified order.
```

Comment:
To be more precise, we can write `remaining keys with non-nil values`.
While such keys are usually considered to be the same as missing, internal representation is a bit different.